### PR TITLE
Implement the classifier, and validation and defaulting engine using CUE

### PR DESCRIPTION
**Pull request by luxas**
This makes the classifier use CUE for all policy-related things. It is good enough for our needs and lets us classify components, set default attributes and labels for components once classified, and enforce validation on fields set in KiCad.

I will still need to rebase this PR slightly.

cc @twelho @chiplet

### CODE REVIEW

File: tools/kicad_rs/sample_policy.cue [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	// Note: "footprint_equals_smd" is a user-defined key and only used for object merging
+	labels: footprint_equals_smd: {
+		key: "footprintLibrary"
+		op:  "Equals"
+		values: ["Capacitor_SMD"]
+	}
+}, _tmpl_resistor & {
+	class: "resistor"
+}, _tmpl_resistor & { // shunt_resistor extends all the properties of resistor
+	class: "shunt_resistor"
```
Comment (twelho) [inline tools/kicad_rs/sample_policy.cue]: There's now identifiers in camel case (`footprintLibrary`) and snake case (`shunt_resistor`) here, do we want to establish some convention for this?

File: tools/kicad_rs/sample_policy.cue [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	attributes: {
+		lt_1: {
+			key: "Tolerance"
+			op:  "Lt"
+			values: [1.1]
+		}
+	}
+}]
+
+// #Policy defines requirement and defaults for all components belonging to a class
```
Comment (twelho) [inline tools/kicad_rs/sample_policy.cue]:
```suggestion
// #Policy defines requirements and defaults for all components belonging to a class
```

File: tools/kicad_rs/sample_policy.cue [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+#Policy: {
+	shunt_resistor: {
+		labels: {
+			extra: shunt: "true"
```
Comment (twelho) [inline tools/kicad_rs/sample_policy.cue]: This "inline" syntax looks interesting, but I guess it's valid CUE

File: tools/kicad_rs/sample_policy.cue [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+			datasheet: "foo"
+		}
+		attributes: Tolerance: {
+			value: <1.1 | string
+			unit:  string | *"Ohm"
+		}
+	}
+	capacitor: attributes: Value: {
+		comment: string | *"This is a capacitor :D!"
```
Comment (twelho) [inline tools/kicad_rs/sample_policy.cue]: Just out of curiosity, is the `|` operator commutative?

File: tools/kicad_rs/sample_policy.cue [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	}
+}
+
+// Common "templates":
+
+_tmpl_resistor: {
```
Comment (twelho) [inline tools/kicad_rs/sample_policy.cue]:
```suggestion
// Common "templates"
_tmpl_resistor: {
```

File: tools/kicad_rs/src/cue/common.cue [hunk 6]
```diff
@@ -0,0 +1,29 @@
+// This file defines common definitions for all the policy-related CUE files
+
+// The most important field of stdin that should be validated is the components map
+// The components map uses the component reference as a key
```
Comment (twelho) [inline tools/kicad_rs/src/cue/common.cue]:
```suggestion
// The most important field of stdin that should be validated is the components map.
// The components map uses the component reference as a key.
```
(tricky to read without periods)

File: tools/kicad_rs/src/cue/map.cue [hunk 7]
```diff
@@ -0,0 +1,34 @@
+// This file "maps" all attributes and labels for a component to a "generated" sub-field
+// where both the class-defined requirements and defaults are merged with the set values of
+// the component
+// This file depends on common.cue
+// TODO: Figure out how to deal with multiple nested schematics
```
Comment (twelho) [inline tools/kicad_rs/src/cue/map.cue]: Nested schematic support is quite important, those are already heavily used in Racklet.

File: tools/kicad_rs/src/cue/policy_schema.cue [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+// The user-defined #Policy object specifies what attributes and labels are mandated
+// for each class, along with any needed defaults
+#Policy: [string]: {
+    attributes: [string]: #Attribute
+    labels: #Labels
+}
+
+// The user-defined #Classifiers object specifies how various attributes are labels
+// can be used to classify a component to belonging to a class.
```
Comment (twelho) [inline tools/kicad_rs/src/cue/policy_schema.cue]: There's something wonky with this sentence, "attributes are labels can be used"?

File: tools/kicad_rs/src/cue/policy_schema.cue [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+#Classifiers: [...#Classifier]
+#Classifier: {
+    class: string
+    labels?: #RequirementMap
+    attributes?: #RequirementMap
+}
+
+// The requirement string can be arbitrarily defined, it's used to be able to merge
+// classes extending each other.
```
Comment (twelho) [inline tools/kicad_rs/src/cue/policy_schema.cue]:
```suggestion
// The requirement string can be arbitrarily defined, it's used to enable
// the merging of classes extending each other
```

File: tools/kicad_rs/src/policy.rs [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+const CUE_POLICY_SCHEMA_BYTES: &'static str = include_str!("cue/policy_schema.cue");
+const CUE_POLICY_SCHEMA_FILE: &'static str = "policy_schema.cue";
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+#[serde(deny_unknown_fields)]
+pub struct ComponentClassifier {
+    // The class that shall be applied to a component matching these requirements
+    pub class: String,
+    // Labels matching
```
Comment (twelho) [inline tools/kicad_rs/src/policy.rs]:
```suggestion
    // Label matching
```

File: tools/kicad_rs/src/policy.rs [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+use std::str;
+use tempfile::tempdir;
+
+const CUE_COMMON_BYTES: &'static str = include_str!("cue/common.cue");
+const CUE_MAP_BYTES: &'static str = include_str!("cue/map.cue");
+const CUE_MAP_FILE: &'static str = "map.cue";
+const CUE_REDUCE_BYTES: &'static str = include_str!("cue/reduce.cue");
+const CUE_REDUCE_FILE: &'static str = "reduce.cue";
+const CUE_POLICY_SCHEMA_BYTES: &'static str = include_str!("cue/policy_schema.cue");
+const CUE_POLICY_SCHEMA_FILE: &'static str = "policy_schema.cue";
```
Comment (twelho) [inline tools/kicad_rs/src/policy.rs]: TIL that one can import a file into the binary at compile time with a single macro.

**Review by twelho — APPROVED** [review]
Very good start! Just a couple of nits/questions (and of course this needs rebasing). The CUE descriptions look quite clear even though I don't have enough experience in the language yet to properly review them.

There was a TODO for supporting hierarchical/nested sheets, so if those aren't supported yet, we need to get them working before this can be utilized in full.

It might also be valuable to build a specialized CUE binary (there was some Go tool to do this IIRC) with the required functionality built in, writing stuff to a temporary directory and execing a binary in $PATH isn't very integrated (but fine for now).

**Comment by twelho** [comment]
This will need some rebasing now that #16 moved the `kicad_rs` files into a Cargo workspace.

File: tools/kicad_rs/sample_policy.cue (hunk 3 above)
Comment (luxas) [reply tools/kicad_rs/sample_policy.cue]: Yes, it is valid CUE, and pretty handy in quite a lot of cases actually.

File: tools/kicad_rs/sample_policy.cue (hunk 4 above)
Comment (luxas) [reply tools/kicad_rs/sample_policy.cue]: IIRC, yes

**Comment by luxas** [comment]
Rebased and ready to merge, pending a re-review @twelho

**Comment by luxas** [comment]
I'll merge this now to unblock @chiplet, I'll fix any feedback in follow-ups in future PRs

**Pull request merged**